src/server/img_storage.rs

```diff
@@ -160,19 +248,31 @@ impl StoreImage {
     /// If you have "Storage Legacy Bucket Writer" previous content is overwritten.
     /// (e.g. set the path to be the SHA1 of the bytes or whatever.)
 
-    pub async fn store(&self, uri: &uri::Uri) -> HandlerResult<StoreResult> {
-        trace!("fetching... {:?}", &uri);
-        /*
-        // Should we preserve the name of the image?
-        let mut hasher = Sha256::new();
-        hasher.update(uri.path().as_bytes());
-        let hash = hex::encode(hasher.finalize().as_slice());
-        */
+    pub async fn store(&self, uri: &uri::Uri) -> HandlerResult<StoredImage> {
+        let (image, content_type) = self.fetch(uri).await?;
+        let metrics = self.validate(uri, &image, &content_type).await?;
+        self.upload(image, &content_type, metrics).await
+    }
 
+    /// Generate a unique hash based on the content of the image
+    pub fn as_hash(&self, source: &Bytes) -> String {
+        let mut hasher = DefaultHasher::new();
+        source.hash(&mut hasher);
+        base64::encode_config(hasher.finish().to_ne_bytes(), base64::URL_SAFE_NO_PAD)
+    }
+
+    /// Fetch the bytes for an image based on a URI
+    pub(crate) async fn fetch(&self, uri: &uri::Uri) -> HandlerResult<(Bytes, String)> {
+        trace!("fetching... {:?}", &uri);
         let res = reqwest::get(&uri.to_string())
```

Comment:
- Let's give `StorageImage` a reference to the `reqwest_client` built at `Server` creation and call reqwest through that instead (to take advantage of the underlying per host pooling)
- Probably worth adding a timeout to the get call like we do for adM
- Definitely want an `.error_for_status()?` for the get too
- Reqwest errors have an `Into<HandlerError>` so map_err shouldn't be necessary (the reqwest error and its traceback should tell us enough about this error)